### PR TITLE
Tags and basic duration aggregation

### PR DESCRIPTION
**Pull request by Erbenos**
next up: link against PCP, clean up copied files

### CODE REVIEW

**Comment by Erbenos** [comment]
I am already linkging against PCP (my fork), ini needs to be separate as pcp doesnt ship the header for it, that should be all files cleaned up as well.

File: src/main.c [hunk 1]
```diff
@@ -27,7 +27,7 @@ int main(int argc, char **argv)
     agent_config* config = (agent_config*) malloc(sizeof(agent_config));
     ALLOC_CHECK(NULL, "Unable to asssign memory for agent config.");
     int config_src_type = argc >= 2 ? READ_FROM_CMD : READ_FROM_FILE;
-    config = read_agent_config(config_src_type, "pcp-statsd-pmda.conf", argc, argv);
+    config = read_agent_config(config_src_type, "config.ini", argc, argv);
```
Comment (lzap) [inline src/main.c]: Extension ini already imply configuration data, I'd go with `statsd-pmda.ini` or something like that.

File: src/main.c [hunk 2]
```diff
@@ -48,13 +48,13 @@ int main(int argc, char **argv)
...
+        die(__FILE__, __LINE__, "Error joining network listener thread.");
     }
     if (pthread_join(datagram_parser, NULL) != 0) {
-        die(__LINE__, "Error joining datagram parser thread.");
+        die(__FILE__, __LINE__, "Error joining datagram parser thread.");
     }
     if (pthread_join(datagram_consumer, NULL) != 0) {
-        die(__LINE__, "Error joining datagram consumer thread.");
+        die(__FILE__, __LINE__, "Error joining datagram consumer thread.");
     }
```
Comment (lzap) [inline src/main.c]: Probably some macro for those __FILE__ and __LINE__ could be appropriate.

File: src/statsd-parsers/basic/basic.c [hunk 3]
```diff
@@ -194,31 +205,45 @@ int parse(char* buffer, statsd_datagram** datagram) {
...
+            snprintf(buffer + current_size, JSON_BUFFER_SIZE - current_size, ",\"%s\":\"%s\"",
+                tags->values[i]->key, tags->values[i]->value);
+        }
+    }
+    current_size = strlen(buffer);
+    buffer[current_size] = '}';
+    char* result = malloc(sizeof(char) * (current_size + 2));
+    ALLOC_CHECK("Unable to allocate memory for tags json.");
+    memcpy(result, buffer, current_size + 2);
+    return result;
```
Comment (lzap) [inline src/statsd-parsers/basic/basic.c]: This one really needs an unit test.

File: src/statsd-parsers/basic/basic.c [hunk 4]
```diff
@@ -194,31 +205,45 @@ int parse(char* buffer, statsd_datagram** datagram) {
...
+    int res = strcmp((*(tag**)x)->key, (*(tag**)y)->key);
+    return res;
+}
+
+/**
+ * Converts tag_collection* struct to JSON string that is sorted by keys
+ */
+char* tag_collection_to_json(tag_collection* tags) {
+    char buffer[JSON_BUFFER_SIZE];
+    memset(buffer, '\0', JSON_BUFFER_SIZE);
```
Comment (lzap) [inline src/statsd-parsers/basic/basic.c]: Is this necessary? You are going to overwrite it anyway and you can append the trailing null characters at the end.

File: src/statsd-parsers/basic/basic.c [hunk 5]
```diff
@@ -194,31 +205,45 @@ int parse(char* buffer, statsd_datagram** datagram) {
...
+    for (i = 0; i < tags->length; i++) {
+        current_size = strlen(buffer);
+        if (i == 0) {
+            snprintf(buffer + current_size, JSON_BUFFER_SIZE - current_size, "\"%s\":\"%s\"",
+                tags->values[i]->key, tags->values[i]->value);
+        } else {
+            snprintf(buffer + current_size, JSON_BUFFER_SIZE - current_size, ",\"%s\":\"%s\"",
+                tags->values[i]->key, tags->values[i]->value);
+        }
+    }
```
Comment (lzap) [inline src/statsd-parsers/basic/basic.c]: This loop must check for buffer overflow every round.

File: src/statsd-parsers/basic/basic.c [hunk 6]
```diff
@@ -194,31 +205,45 @@ int parse(char* buffer, statsd_datagram** datagram) {
...
+    current_size = strlen(buffer);
+    buffer[current_size] = '}';
```
Comment (lzap) [inline src/statsd-parsers/basic/basic.c]: One extra check before the final } (and null char) are appended is needed too.

**Comment by lzap** [comment]
Oh you meant to fix trafix in this PR or another PR? PCP is missing there.

Note that Travis uses Ubuntu in containers, you can try to install "apt-get install pcp-dev" and that shall work. If not, then wget pcp source and compile it from scratch on travis. This is always painful to set, if you end up spending too much time on this, just remove travis from here - we are going to eventually move to PCP anyway.

File: src/statsd-parsers/basic/basic.c (hunk 5 above)
Comment (natoscott) [reply src/statsd-parsers/basic/basic.c]: Please use pmsprintf (from libpcp) in place of sprintf/snprintf everywhere - it guarantees null-termination and all PCP code has been audited & switched over (this was a requirement from the Ubuntu security teams review of PCP a couple years ago).

File: src/main.c (hunk 2 above)
Comment (Erbenos) [reply src/main.c]: Fixed by a34f450

File: src/main.c (hunk 1 above)
Comment (Erbenos) [reply src/main.c]: Fixed by 45c6bf7

File: src/config-reader/config-reader.h [hunk 7]
```diff
@@ -1,22 +1,33 @@
...
-#ifndef CONFIG_READER_
-#define CONFIG_READER_
-
 const int READ_FROM_FILE;
 const int READ_FROM_CMD;
 
+typedef enum PARSER_TYPE {
+    TRIVIAL = 0b0,
+    RAGEL = 0b1
+} PARSER_TYPE;
```
[the remark this answers is not in the document]
Comment (Erbenos) [reply src/config-reader/config-reader.h]: TRIVIAL => BASIC in 0554de0

File: src/statsd-parsers/basic/basic.c (hunk 4 above)
Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: Memseting buffers to \0 removed in 726522d

File: src/statsd-parsers/basic/basic.c (hunk 5 above)
Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: Replaced sprintf/snprinf with pmsprintf in 2ff5e7a

Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: @lzap the buffer never overflows as pmsprintf/snprintf is never writing more characters then is current remaining buffer size

File: src/statsd-parsers/basic/basic.c (hunk 6 above)
Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: Fixed in b92b51c

File: src/consumers/consumers.h [hunk 8]
```diff
@@ -106,9 +113,59 @@ metric_metadata* create_metric_meta(statsd_datagram* datagram);
...
+
+/**
+ * Gets duration values meta data from given collection, as a sideeffect it sorts the values
+ * @arg collection - Target collection
+ * @arg out - Placeholder for data population
+ * @return 1 on success
+ */
+int get_bduration_values_meta(bduration_collection* collection, duration_values_meta* out);
 
 #endif
```
[the remark this answers is not in the document]
Comment (Erbenos) [reply src/consumers/consumers.h]: I made these bduration functions static in dfd284f

"This is by the way probably better term than consumer."
I dont understand this.

File: src/statsd-parsers/basic/basic.c (hunk 3 above)
Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: @lzap @natoscott 
What do you think is most fitting approach to this?
Make separate Makefile rule with -D flag with tested unit as an entry point with main inside NDEBUG macro flag?

**Comment by Erbenos** [comment]
I added PCP to Travis - building from source, wasn't that complicated as I was doing basically the same for chan and HdrHistogram_c already.

File: src/statsd-parsers/basic/basic.c (hunk 3 above)
Comment (Erbenos) [reply src/statsd-parsers/basic/basic.c]: Test in a9bb2d2

**Pull request merged**